sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamWindowRel.java

```diff
@@ -0,0 +1,270 @@
+
+    private Schema outputSchema;
+    private List<FieldAggregation> aggFields;
+
+    public Transform(Schema s, List<FieldAggregation> af) {
+      this.outputSchema = s;
+      this.aggFields = af;
+    }
+
+    @Override
+    public PCollection<Row> expand(PCollectionList<Row> input) {
+      PCollection<Row> r = input.get(0);
+      for (FieldAggregation af : aggFields) {
+        org.apache.beam.sdk.schemas.transforms.Group.ByFields<Row> myg =
+            org.apache.beam.sdk.schemas.transforms.Group.byFieldIds(af.partitionKeys);
+        r = r.apply("partitionBy", myg);
+        r = r.apply("orderBy", ParDo.of(sortPartition(af))).setRowSchema(r.getSchema());
+        r = r.apply("aggCall", ParDo.of(aggField(outputSchema, af))).setRowSchema(outputSchema);
+      }
+      return r;
+    }
+  }
+
+  private static DoFn<Row, Row> aggField(
+      final Schema outputSchema, final FieldAggregation fieldAgg) {
+    return new DoFn<Row, Row>() {
+      @ProcessElement
+      public void processElement(
+          @Element Row inputPartition, OutputReceiver<Row> out, ProcessContext c) {
+        Collection<Row> inputPartitions = inputPartition.getArray(1); // 1 -> value
+        List<Row> sortedRowsAsList = new ArrayList<Row>(inputPartitions);
+        for (int idx = 0; idx < sortedRowsAsList.size(); idx++) {
+          int lowerIndex = idx - fieldAgg.lowerLimit;
+          int upperIndex = idx + fieldAgg.upperLimit + 1;
+          lowerIndex = lowerIndex < 0 ? 0 : lowerIndex;
+          upperIndex = upperIndex > sortedRowsAsList.size() ? sortedRowsAsList.size() : upperIndex;
+          List<Row> aggRange = sortedRowsAsList.subList(lowerIndex, upperIndex);
+
+          // Just concept-proof
+          // Assume that aggFun = SUM
+          // Assume dataType = INTEGER
+          final Combine.CombineFn<Integer, int[], Integer> aggFunction =
+              (Combine.CombineFn<Integer, int[], Integer>)
+                  BeamBuiltinAggregations.create("SUM", Schema.FieldType.INT32);
+          int[] aggAccumulator = aggFunction.createAccumulator();
+
+          // Assume a simple expression within SUM($aUniqueDirectField)
+          final int aggFieldIndex = fieldAgg.inputFields.get(0);
+
+          for (Row aggRow : aggRange) {
+            Integer valueToAgg = aggRow.getInt32(aggFieldIndex);
+            aggFunction.addInput(aggAccumulator, valueToAgg);
+          }
+          Integer aggOutput = aggFunction.extractOutput(aggAccumulator);
+          List<Object> fieldValues =
+              Lists.newArrayListWithCapacity(sortedRowsAsList.get(idx).getFieldCount());
+          fieldValues.addAll(sortedRowsAsList.get(idx).getValues());
+          fieldValues.add(aggOutput);
+          Row ou = Row.withSchema(outputSchema).addValues(fieldValues).build();
+          out.output(ou);
+        }
+      }
+    };
+  }
+
+  private static DoFn<Row, Row> sortPartition(final FieldAggregation fieldAgg) {
```

Comment:
I am wondering if this transform can be reused? https://github.com/apache/beam/blob/master/sdks/java/extensions/sorter/src/main/java/org/apache/beam/sdk/extensions/sorter/SortValues.java

Reply:
It does not bad to start from this ParDo though. I posted that transform just for curiosity. Not required to use SortValues at this moment.